Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerHandler.swift

```diff
@@ -616,25 +583,54 @@ internal final class AsyncServerHandler<
 
   @Sendable
   @inlinable
-  internal func interceptResponseMessage(_ response: Response, compression: Compression) {
+  internal func interceptResponseMessages(_ messages: Deque<(Response, Compression)>) {
     if self.eventLoop.inEventLoop {
-      self._interceptResponseMessage(response, compression: compression)
+      for message in messages {
+        self._interceptResponseMessage(message.0, compression: message.1)
+      }
     } else {
       self.eventLoop.execute {
-        self._interceptResponseMessage(response, compression: compression)
+        for message in messages {
+          self._interceptResponseMessage(message.0, compression: message.1)
+        }
       }
     }
   }
 
   @inlinable
-  internal func _interceptResponseStatus(_ status: GRPCStatus) {
+  internal func _interceptTermination(_ error: Error?) {
     self.eventLoop.assertInEventLoop()
 
+    let processedError: Error?
+    if let thrownStatus = error as? GRPCStatus, thrownStatus.isOk {
+      processedError = GRPCStatus(
+        code: .unknown,
+        message: "Handler threw error with status code 'ok'."
+      )
+    } else {
+      processedError = error
+    }
+
     switch self.handlerStateMachine.sendStatus() {
-    case let .intercept(_, trailers):
+    case let .intercept(requestHeaders, trailers):
+      let status: GRPCStatus
+      let processedTrailers: HPACKHeaders
+
+      if processedError != nil {
+        (status, processedTrailers) = ServerErrorProcessor.processObserverError(
+          processedError!,
```

Comment:
We can avoid the IOU here:

```suggestion
      if let processedError = processedError {
        (status, processedTrailers) = ServerErrorProcessor.processObserverError(
          processedError,
```